File: lib/src/internal/headers.dart
part of frugal;

var _encoder = new Utf8Encoder();
var _decoder = new Utf8Decoder();

/**
 * This is an internal-only class. Don't use it!
 */
class Headers {
  static const _V0 = 0x00;

  /// Encode the headers
  static Uint8List encode(Map<String, String> headers) {
    var size = 0;
    // Get total frame size headers
    if (headers != null && headers.length > 0) {
      for (var name in headers.keys) {
        // 4 bytes each for name, value length
        size += 8 + name.length + headers[name].length;
      }
    }

    // Header buff = [version (1 byte), size (4 bytes), headers (size bytes)]
    var buff = new Uint8List(5 + size);

    // Write version
    buff[0] = _V0;

    // Write size
    _writeInt(size, buff, 1);

    // Write headers
    if (headers != null && headers.length > 0) {
      var i = 5;
      for (var name in headers.keys) {
        // Write name length
        _writeInt(name.length, buff, i);
        i += 4;
        // Write name
        _writeString(name, buff, i);
        i += name.length;

        // Write value length
        var value = headers[name];
        _writeInt(value.length, buff, i);
        i += 4;
        _writeString(value, buff, i);
        i += value.length;
      }
    }
    return buff;
  }

  /// Reads the headers from a TTransort
  static Map<String, String> read(TTransport transport) {
    // Buffer version
    var buff = new Uint8List(5);
    transport.readAll(buff, 0, 1);

    // Support more versions when available
    if (buff[0] != _V0) {
      throw new FError.withMessage("unsupported header version ${buff[0]}");
    }

    // Read size
    transport.readAll(buff, 1, 4);
    var size = _readInt(buff, 1);

    // Read the rest of the header bytes into a buffer
    buff = new Uint8List(size);
    transport.readAll(buff, 0, size);

    return _readPairs(buff, 0, size);
  }

  /// Returns the headers from Frugal frame
  static Map<String, String> decodeFromFrame(Uint8List frame) {
    if (frame.length < 5) {
      throw new FError.withMessage("invalid frame size ${frame.length}");
    }

    // Support more versions when available
    if (frame[0] != _V0) {
      throw new FError.withMessage("unsupported header version ${frame[0]}");
    }

    return _readPairs(frame, 5, _readInt(frame, 1) + 5);
  }

  static Map<String, String> _readPairs(Uint8List buff, int start, int end) {
    var headers = {};
    for (var i = start; i < end; i) {
      // Read header name
      var nameSize = _readInt(buff, i);
      i += 4;
      if (i > end || i + nameSize > end) {
        throw new FError.withMessage("invalid protocol header name");
      }
      var name = _decoder.convert(buff, i, i + nameSize);
      i += nameSize;

      // Read header value
      var valueSize = _readInt(buff, i);
      i += 4;
      if (i > end || i + valueSize > end) {
        throw new FError.withMessage("invalid protocol header value");
      }
      var value = _decoder.convert(buff, i, i + valueSize);
      i += valueSize;

      // Set the pair
      headers[name] = value;
    }
    return headers;
  }

  static int _readInt(Uint8List buff, int i) {
    return ((buff[i] & 0xff) << 24) |
        ((buff[i + 1] & 0xff) << 16) |
        ((buff[i + 2] & 0xff) << 8) |
        (buff[i + 3] & 0xff);
  }

  static void _writeInt(int i, Uint8List buff, int i1) {
    buff[i1] = (0xff & (i >> 24));
    buff[i1 + 1] = (0xff & (i >> 16));
    buff[i1 + 2] = (0xff & (i >> 8));
    buff[i1 + 3] = (0xff & (i));
  }

  static void _writeString(String s, Uint8List buff, int i) {
    var strBytes = _encoder.convert(s);
    buff.setRange(i, i + strBytes.length, strBytes);
  }
}

File: lib/java/src/main/java/com/workiva/frugal/internal/Headers.java
package com.workiva.frugal.internal;

import com.workiva.frugal.exception.FProtocolException;
import com.workiva.frugal.util.Pair;
import com.workiva.frugal.util.ProtocolUtils;
import org.apache.thrift.TException;
import org.apache.thrift.protocol.TProtocolException;
import org.apache.thrift.transport.TTransport;
import org.apache.thrift.transport.TTransportException;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * This is an internal-only class. Don't use it!
 */
public class Headers {
    // Version 0
    private static final byte V0 = 0x00;

    public static byte[] encode(Map<String, String> headers) throws TException {
        int size = 0;
        if (headers == null) {
            headers = new HashMap<>();
        }

        List<Pair<byte[], byte[]>> utf8Headers = new ArrayList<>();
        // Get total frame size headers
        for (Map.Entry<String, String> pair : headers.entrySet()) {
            byte[] keyBytes = ProtocolUtils.encodeString(pair.getKey());
            byte[] valueBytes = ProtocolUtils.encodeString(pair.getValue());
            size += 8 + keyBytes.length + valueBytes.length;
            utf8Headers.add(Pair.of(keyBytes, valueBytes));
        }

        byte[] buff = new byte[size + 5];

        // Write version
        buff[0] = V0;

        // Write size
        ProtocolUtils.writeInt(size, buff, 1);

        int i = 5;
        // Write headers
        for (Pair<byte[], byte[]> pair : utf8Headers) {
            // Write key
            byte[] key = pair.getLeft();
            ProtocolUtils.writeInt(key.length, buff, i);
            i += 4;
            ProtocolUtils.writeStringBytes(key, buff, i);
            i += key.length;

            // Write value
            byte[] value = pair.getRight();
            ProtocolUtils.writeInt(value.length, buff, i);
            i += 4;
            ProtocolUtils.writeStringBytes(value, buff, i);
            i += value.length;
        }
        return buff;
    }

    public static Map<String, String> read(TTransport transport) throws TException {
        byte[] buff = new byte[5];

        // Read version
        transport.readAll(buff, 0, 1);

        // Support more versions when available
        if (buff[0] != V0) {
            throw new FProtocolException(FProtocolException.BAD_VERSION, "unsupported header version " + buff[0]);
        }

        // Read size
        transport.readAll(buff, 1, 4);
        int size = ProtocolUtils.readInt(buff, 1);
        buff = new byte[size];
        transport.readAll(buff, 0, size);

        return readPairs(buff, 0, size);
    }

    public static Map<String, String> decodeFromFrame(byte[] frame) throws TException {
        if (frame.length < 5) {
            throw new FProtocolException(FProtocolException.INVALID_DATA, "invalid frame size " + frame.length);
        }

        // Support more versions when available
        if (frame[0] != V0) {
            throw new FProtocolException(FProtocolException.BAD_VERSION, "unsupported header version " + frame[0]);
        }

        return readPairs(frame, 5, ProtocolUtils.readInt(frame, 1) + 5);
    }

    private static Map<String, String> readPairs(byte[] buff, int start, int end) throws TException {
        Map<String, String> headers = new HashMap<>();
        int i = start;
        while (i < end) {
            try {
                // Read header name
                int nameSize = ProtocolUtils.readInt(buff, i);
                i += 4;
                if (i > end || i + nameSize > end) {
                    throw new FProtocolException(TProtocolException.INVALID_DATA, "invalid protocol header name");
                }
                byte[] nameBuff = Arrays.copyOfRange(buff, i, nameSize + i);
                i += nameSize;
                String name = new String(nameBuff, "UTF-8");

                // Read header value
                int valueSize = ProtocolUtils.readInt(buff, i);
                i += 4;
                if (i > end || i + valueSize > end) {
                    throw new FProtocolException(FProtocolException.INVALID_DATA, "invalid protocol header value");
                }
                byte[] valueBuff = Arrays.copyOfRange(buff, i, valueSize + i);
                i += valueSize;
                String value = new String(valueBuff, "UTF-8");

                headers.put(name, value);
            } catch (IOException e) {
                throw new TTransportException(FProtocolException.UNKNOWN, "could not read header bytes", e);
            }
        }
        return headers;
    }

}
